syntax = "proto3";

// https://clics.ecs.baylor.edu/index.php/Contest_API#Contests

package proto.clics;
option java_package = "edu.clics.proto";
option java_outer_classname = "ClicsProto";

import "google/protobuf/timestamp.proto";
import "google/protobuf/duration.proto";
import "api/clics/proto/annotations.proto";

// Endpoint: none. Used for storage.

message ClicsContest {
  Contest contest = 1;
  ContestState state = 2;
  repeated ScoreboardRow scoreboard = 3;
  repeated EventFeedItem event_feed = 4 [(js.serialized_name) = "event-feed"];

  map<string, JudgementType> judgement_types = 5 [(js.serialized_name) = "judgement-types", json_name = 'judgement-types'];
  map<string, Language> languages = 6;
  map<string, Problem> problems = 7;
  map<string, Group> groups = 8;
  map<string, Organization> organizations = 9;
  map<string, Team> teams = 10;
  map<string, TeamMember> team_members = 11 [(js.serialized_name) = "team-members", json_name = 'team-members'];
  map<string, Submission> submissions = 12;
  map<string, Judgement> judgements = 13;
  map<string, Run> runs = 14;
  map<string, Clarification> clarifications = 15;
  map<string, Award> awards = 16;
}

// Endpoint: GET /contests => Contest[]
// Endpoint: GET /contests/{} => Contest

message Contest {
  string id = 1;
  string name = 2;
  string formal_name = 3 [json_name = 'formal_name'];

  google.protobuf.Timestamp start_time = 4 [json_name = 'start_time'];
  google.protobuf.Duration contest_duration = 5 [(js.serialized_name) = "duration", json_name = 'duration'];
  google.protobuf.Duration scoreboard_freeze_duration = 6 [json_name = 'scoreboard_freeze_duration'];

  sint64 penalty_time = 7 [json_name = 'penalty_time'];
  repeated Media banners = 8 [(js.serialized_name) = "banner", json_name = 'banner'];
  repeated Media logos = 9 [(js.serialized_name) = "logo", json_name = 'logos'];
}

// Endpoint: GET /contests/{}/judgement-types => JudgementType[]

message JudgementType {
  string id = 1;
  string name = 2;
  bool penalty = 3;
  bool solved = 4;
}

// Endpoint: GET /contests/{}/languages => Language[]
// Endpoint: GET /contests/{}/languages/{} => Language

message Language {
  string id = 1;
  string name = 2;
}

// Endpoint: GET /contests/{}/problems => Problem[]
// Endpoint: GET /contests/{}/problems/{} => Problem

message Problem {
  string id = 1;
  string label = 2;
  string name = 3;
  sint32 ordinal = 4;
  string rgb = 5;
  string color = 6;
  double time_limit = 7 [json_name = 'time_limit'];
  uint32 test_data_count = 8 [json_name = 'test_data_count'];
}

// Endpoint: GET /contests/{}/groups => Group[]
// Endpoint: GET /contests/{}/groups/{} => Group

message Group {
  string id = 1;
  string icpc_id = 2 [json_name = 'icpc_id'];
  string name = 3;
  string type = 4;
  bool hidden = 5;
}

// Endpoint: GET /contests/{}/organizations => Organization[]
// Endpoint: GET /contests/{}/organizations/{} => Organization

message Organization {
  string id = 1;
  string icpc_id = 2 [json_name = 'icpc_id'];
  string name = 3;
  string formal_name = 4 [json_name = 'formal_name'];
  string country = 5;
  string url = 6;
  string social_media_tag = 7 [(js.serialized_name) = "twitter_hashtag", json_name = 'twitter_hashtag'];
  OrganizationLocation location = 8;
  repeated Media logos = 9 [(js.serialized_name) = "logo", json_name = 'logo'];
}

message OrganizationLocation {
  double latitude = 1;
  double longitude = 2;
}

// Endpoint: GET /contests/{}/teams => Team[]
// Endpoint: GET /contests/{}/teams/{} => Team

message Team {
  string id = 1;
  string icpc_id = 2 [json_name = 'icpc_id'];
  string name = 3;
  string organization_id = 4 [json_name = 'organization_id'];
  repeated string group_ids = 5 [(js.serialized_name) = "group_ids", json_name = 'group_ids'];
  TeamLocation location = 6;
  repeated Media photos = 7 [(js.serialized_name) = "photo", json_name = 'photo'];
  repeated Media videos = 8 [(js.serialized_name) = "video", json_name = 'video'];
  repeated Media desktop_stream = 9 [(js.serialized_name) = "desktop", json_name = 'desktop'];
  repeated Media webcam_stream = 10 [(js.serialized_name) = "webcam", json_name = 'webcam'];
}

message TeamLocation {
  double x = 1; // metres
  double y = 2; // metres
  double rotation = 3; // degrees
}

// Endpoint: GET /contests/{}/team-members => TeamMember[]
// Endpoint: GET /contests/{}/team-members/{} => TeamMember

message TeamMember {
  string id = 1;
  string icpc_id = 2 [json_name = 'icpc_id'];
  string team_id = 3 [json_name = 'team_id'];
  string first_name = 4 [json_name = 'first_name'];
  string last_name = 5 [json_name = 'last_name'];
  string sex = 6;
  string role = 7;
  repeated Media photos = 8 [(js.serialized_name) = "photo", json_name = 'photo'];
}

// Endpoint: /contests/{}/state => ContestState

message ContestState {
  google.protobuf.Timestamp started = 1;
  google.protobuf.Timestamp frozen = 2;
  google.protobuf.Timestamp ended = 3;
  google.protobuf.Timestamp thawed = 4;
  google.protobuf.Timestamp finalized = 5;
}

// Endpoint: /contests/{}/submissions => Submission[]
// Endpoint: /contests/{}/submissions/{} => Submission

message Submission {
  string id = 1;
  string language_id = 2 [json_name = 'language_id'];
  string problem_id = 3 [json_name = 'problem_id'];
  string team_id = 4 [json_name = 'team_id'];
  google.protobuf.Timestamp time = 5;
  google.protobuf.Duration contest_time = 6 [json_name = 'contest_time'];
  string entry_point = 7 [json_name = 'entry_point'];
  repeated Media files = 8;
  repeated Media reactionVideo = 9 [(js.serialized_name) = "reaction", json_name = 'reaction'];
}

// Endpoint: /contests/{}/judgements => Judgement[]
// Endpoint: /contests/{}/judgements/{} => Judgement

message Judgement {
  string id = 1;
  string submission_id = 2 [json_name = 'submission_id'];
  string judgement_type_id = 3 [json_name = 'judgement_type_id'];
  google.protobuf.Timestamp start_time = 4 [json_name = 'start_time'];
  google.protobuf.Duration start_contest_time = 5 [json_name = 'contest_time'];
  google.protobuf.Timestamp end_time = 6 [json_name = 'end_time'];
  google.protobuf.Duration end_contest_time = 7 [json_name = 'end_contest_time'];
  double max_run_time = 8;
}

// Endpoint: /contests/{}/runs => Run[]
// Endpoint: /contests/{}/runs/{} => Run

message Run {
  string id = 1;
  string judgement_id = 2 [json_name = 'judgement_id'];
  uint32 ordinal = 3;
  string judgement_type_id = 4 [json_name = 'type_id'];
  google.protobuf.Timestamp time = 5;
  google.protobuf.Duration contest_time = 6 [json_name = 'contest_time'];
  double run_time = 7 [json_name = 'run_time'];
}

// Endpoint: /contests/{}/clarifications => Clarification[]
// Endpoint: /contests/{}/clarifications/{} => Clarification

message Clarification {
  string id = 1;
  string from_team_id = 2 [json_name = 'from_team_id'];
  string to_team_id = 3 [json_name = 'to_team_id'];
  string reply_to_id = 4 [json_name = 'reply_to_id'];
  string problem_id = 5 [json_name = 'problem_id'];
  string text = 6;
  google.protobuf.Timestamp time = 7;
  google.protobuf.Duration contest_time = 8 [json_name = 'contest_time'];
}

// Endpoint: /contests/{}/awards/ => Award[]
// Endpoint: /contests/{}/awards/{} => Award

message Award {
  string id = 1;
  string citation = 2;
  repeated string team_ids = 3 [json_name = 'team_ids'];
}

// Endpoint: /contests/{}/scoreboard => ScoreboardRow[] (2018)
//                                   => Scoreboard (2019)

message Scoreboard {
  string event_id = 1;
  google.protobuf.Timestamp time = 2 [json_name = 'time'];
  google.protobuf.Duration contest_time = 3 [json_name = 'contest_time'];
  // ContestState state = 4;
  repeated ScoreboardRow rows = 5;
}

message ScoreboardRow {
  uint64 rank = 1;
  string team_id = 2 [json_name = 'team_id'];
  ScoreboardScore score = 3;
  repeated ScoreboardProblem problems = 4;
}

message ScoreboardScore  {
  uint32 num_solved = 1 [json_name = 'num_solved'];
  sint64 total_time = 2 [json_name = 'total_time'];
}

message ScoreboardProblem {
  string problem_id = 1 [json_name = 'problem_id'];
  uint32 num_judged = 2 [json_name = 'num_judged'];
  uint32 num_pending = 3 [json_name = 'num_pending'];
  bool solved = 4;
  sint64 time = 5;
}

// Endpoint: /contests/{}/event-feed => EventFeedItem[]

// TODO: implement event feed. Needs extra parse rules because of bizarre stringly-typed format.

message EventFeedItem {
  string id = 1;
  string type = 2;
  string op = 3;
  // PolymorphicMess data = 4;
}

// Basic types

message Media {
  string href = 1;
  string mime = 2;
  uint32 width = 3;
  uint32 height = 4;
}


